### PR TITLE
feat: `PetEnhancement` graphQL query

### PR DESCRIPTION
**Pull request by boscohyun**
- Add `ActionQueryFieldType`
- Add `PetEnhancementFieldType`
- Apply `petEnhancement` action query

### Doc
image

### Query

```graphql
query {
  actionQuery {
    petEnhancement(
      avatarAddress: "0xe56d432da2032f6C851943b76C4a41815baaBB54",
      petId: 1,
      targetLevel: 1)
  }
}
```

```json
{
  "data": {
    "actionQuery": {
      "petEnhancement": "6475373a747970655f69647531353a7065745f656e68616e63656d656e7475363a76616c7565736475313a6132303ae56d432da2032f6c851943b76c4a41815baabb5475323a696431363acd57608e9f06b34f9f279651d9ba9aa675313a7075313a3175313a7475313a316565"
    }
  },
  "extensions": {}
}
```

```python
>>> bytes.fromhex("6475373a747970655f69647531353a7065745f656e68616e63656d656e7475363a76616c7565736475313a6132303ae56d432da2032f6c851943b76c4a41815baabb5475323a696431363acd57608e9f06b34f9f279651d9ba9aa675313a7075313a3175313a7475313a316565")
b"du7:type_idu15:pet_enhancementu6:valuesdu1:a20:\xe5mC-\xa2\x03/l\x85\x19C\xb7lJA\x81[\xaa\xbbTu2:id16:\xcdW`\x8e\x9f\x06\xb3O\x9f'\x96Q\xd9\xba\x9a\xa6u1:pu1:1u1:tu1:1ee"
>>> loads(_)
{'type_id': 'pet_enhancement', 'values': {'a': b'\xe5mC-\xa2\x03/l\x85\x19C\xb7lJA\x81[\xaa\xbbT', 'id': b"\xcdW`\x8e\x9f\x06\xb3O\x9f'\x96Q\xd9\xba\x9a\xa6", 'p': '1', 't': '1'}}
```

### CODE REVIEW

File: NineChronicles.Headless.Tests/GraphTypes/ActionQueryFieldTypes/ActionQueryFieldTypeTest.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    {
+        private static Codec _codec;
+
+        public ActionQueryFieldTypeTest()
+        {
+            _codec = new Codec();
+        }
+
+        [Theory]
+        [InlineData(null, null, false, null)]
```
[the remark this answers is not in the document]
Comment (boscohyun) [reply NineChronicles.Headless.Tests/GraphTypes/ActionQueryFieldTypes/ActionQueryFieldTypeTest.cs:21]: It cannot be.😅

File: NineChronicles.Headless/GraphTypes/ActionQuery.cs [hunk 2]
```diff
@@ -545,6 +546,8 @@ public ActionQuery(StandaloneContext standaloneContext)
             RegisterRapidCombination();
             RegisterCombinationConsumable();
 
+            base.AddField(new PetEnhancementFieldType(Encode));
```
Comment (boscohyun) [inline NineChronicles.Headless/GraphTypes/ActionQuery.cs:549]: @moreal This line calls the parent method. And this line is in the constructor of `ActionQuery`. So I need to turn off the warning for 'SA1100'.

Comment (boscohyun) [reply NineChronicles.Headless/GraphTypes/ActionQuery.cs:549]: If I change this line like below, another warning was occurred.

```suggestion
            AddField(new PetEnhancementFieldType(Encode));
```
 "Virtual member call in constructor"

File: NineChronicles.Headless.Tests/GraphTypes/ActionQueryFieldTypes/ActionQueryFieldTypeTest.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (boscohyun) [reply NineChronicles.Headless.Tests/GraphTypes/ActionQueryFieldTypes/ActionQueryFieldTypeTest.cs]: @moreal I fixed it.

File: NineChronicles.Headless/GraphTypes/ActionQuery.cs (hunk 2 above)
Comment (moreal) [reply NineChronicles.Headless/GraphTypes/ActionQuery.cs:555]: When I reverted it, it seems worked to me. You can see the changes and the build result at https://github.com/moreal/NineChronicles.Headless/commit/e187a82f8ed54a61adb4e8bf63a0dbe5402a1ae3 commit. Is there something I missed? 🙏🏻

Comment (boscohyun) [reply NineChronicles.Headless/GraphTypes/ActionQuery.cs:555]: @moreal That's little bit weird.🙃 I'll retry now.

Comment (boscohyun) [reply NineChronicles.Headless/GraphTypes/ActionQuery.cs]: @moreal The warning message above comes from `ReSharper` linter.. I reverted it. Thanks!

File: NineChronicles.Headless/GraphTypes/ActionQueryFieldTypes/PetEnhancementFieldType.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            var avatarAddr = context.GetArgument<Address>("avatarAddress");
+            if (avatarAddr == default)
+            {
+                throw new ExecutionError("Invalid avatarAddress.");
+            }
+
+            var petId = context.GetArgument<int>("petId");
+            if (petId < 0)
+            {
+                throw new ExecutionError("Invalid petId.");
```
Comment (moreal) [inline NineChronicles.Headless/GraphTypes/ActionQueryFieldTypes/PetEnhancementFieldType.cs:48]:
```suggestion
                throw new ExecutionError("Invalid petId. petId must be positive number.");
```

File: NineChronicles.Headless/GraphTypes/ActionQueryFieldTypes/PetEnhancementFieldType.cs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            }
+
+            var targetLevel = context.GetArgument<int>("targetLevel");
+            if (targetLevel < 1 ||
+                targetLevel == int.MaxValue)
```
Comment (moreal) [inline NineChronicles.Headless/GraphTypes/ActionQueryFieldTypes/PetEnhancementFieldType.cs:53]:
```
targetLevel == int.MaxValue
```

I'm not sure why this condition exists when looking the `PetEnhancement` action 🤔

Comment (boscohyun) [reply NineChronicles.Headless/GraphTypes/ActionQueryFieldTypes/PetEnhancementFieldType.cs:53]: Thanks that is my mistake. I thought a stack overflow case but it's not happen with `targetLevel == int.MaxValue`. 😵‍💫

File: NineChronicles.Headless/GraphTypes/ActionQueryFieldTypes/PetEnhancementFieldType.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            {
+                throw new ExecutionError(
+                    "Invalid petId. petId must greater than or equal to 0");
+            }
+
+            var targetLevel = context.GetArgument<int>("targetLevel");
+            if (targetLevel < 1)
+            {
+                throw new ExecutionError(
+                    "Invalid targetLevel. targetLevel must greater than or equal to 1");
```
Comment (moreal) [inline NineChronicles.Headless/GraphTypes/ActionQueryFieldTypes/PetEnhancementFieldType.cs:56]:
```suggestion
                    "Invalid targetLevel. targetLevel must be greater than or equal to 1");
```

Or it is just a preference:

```suggestion
                    "Invalid targetLevel. targetLevel must be greater than 0");
```

Comment (boscohyun) [reply NineChronicles.Headless/GraphTypes/ActionQueryFieldTypes/PetEnhancementFieldType.cs]: I applied it.

**Review by moreal — APPROVED** [review]

**Pull request merged**